onnxruntime/test/global_thread_pools/test_inference.cc

```diff
@@ -0,0 +1,222 @@
+  size_t input_tensor_size = 224 * 224 * 3;
+  input.values.resize(input_tensor_size);
+  auto& input_tensor_values = input.values;
+  for (unsigned int i = 0; i < input_tensor_size; i++)
+    input_tensor_values[i] = (float)i / (input_tensor_size + 1);
+
+  // prepare expected inputs and outputs
+  expected_dims_y = {1, 1000, 1, 1};
+  // For this test I'm checking for the first 5 values only since the global thread pool change
+  // doesn't affect the core op functionality
+  expected_values_y = {0.000045f, 0.003846f, 0.000125f, 0.001180f, 0.001317f};
+
+  output_name = "softmaxout_1";
+}
+
+// All tests below use global threadpools
+
+// Test 1
+// run inference on a model using just 1 session
+TEST_P(CApiTestGlobalThreadPoolsWithProvider, simple) {
+  // prepare inputs/outputs
+  std::vector<Input> inputs;
+  std::vector<int64_t> expected_dims_y;
+  std::vector<float> expected_values_y;
+  std::string output_name;
+  GetInputsAndExpectedOutputs(inputs, expected_dims_y, expected_values_y, output_name);
+
+  // create session
+  Ort::Session session = GetSessionObj<PATH_TYPE, float>(*ort_env, MODEL_URI, GetParam());
+
+  // run session
+  if (session) {
+    TestInference<PATH_TYPE, float>(session, inputs, output_name.c_str(), expected_dims_y, expected_values_y);
+  }
+}
+
+// Test 2
+// run inference on the same model using 2 sessions
+// destruct the 2 sessions only at the end
+TEST_P(CApiTestGlobalThreadPoolsWithProvider, simple2) {
+  // prepare inputs/outputs
+  std::vector<Input> inputs;
+  std::vector<int64_t> expected_dims_y;
+  std::vector<float> expected_values_y;
+  std::string output_name;
+  GetInputsAndExpectedOutputs(inputs, expected_dims_y, expected_values_y, output_name);
+
+  // create sessions
+  Ort::Session session1 = GetSessionObj<PATH_TYPE, float>(*ort_env, MODEL_URI, GetParam());
+  Ort::Session session2 = GetSessionObj<PATH_TYPE, float>(*ort_env, MODEL_URI, GetParam());
+
+  // run session
+  if (session1 && session2) {
+    TestInference<PATH_TYPE, float>(session1, inputs, output_name.c_str(), expected_dims_y, expected_values_y);
+    TestInference<PATH_TYPE, float>(session2, inputs, output_name.c_str(), expected_dims_y, expected_values_y);
+  }
+}
+
+// Test 3
+// run inference on the same model using 2 sessions
+// one after another destructing first session first
+// followed by second session
+TEST_P(CApiTestGlobalThreadPoolsWithProvider, simple3) {
+  // prepare inputs/outputs
+  std::vector<Input> inputs;
+  std::vector<int64_t> expected_dims_y;
+  std::vector<float> expected_values_y;
+  std::string output_name;
+  GetInputsAndExpectedOutputs(inputs, expected_dims_y, expected_values_y, output_name);
+
+  // first session
+  {
+    // create session
+    Ort::Session session = GetSessionObj<PATH_TYPE, float>(*ort_env, MODEL_URI, GetParam());
+
+    // run session
+    if (session) {
+      TestInference<PATH_TYPE, float>(session, inputs, output_name.c_str(), expected_dims_y, expected_values_y);
```

Comment:
Should these use ORT_ENFORCE(session)? Or at least provide some output on the 'else' path if a session wasn't created.

Reply:
It's not really an error if the session ptr is null. It just means that the specific EP is not getting tested.